jupyter_kernel_mgmt/tests/test_kernelapp.py

```diff
@@ -38,17 +39,28 @@ def test_kernelapp_lifecycle(setup_env):
                                  .format(WAIT_TIME))
 
         # Connection file should be there by now
+        # TODO: now there are 2 connection files (1 to launch the kernel, 1 to
+        # advertise it). Clean this up when we have a better solution.
```

Comment:
Yeah, I agree this is an issue.  One approach would be to always use the kernel-id then kernelapp could easily discover the file by composing the name.  Since the launcher is what creates the file, we'd need to convey the kernel-id to KernelManager's initializer.

In the meantime, I was thinking we could use a prefix of `kernelapp_` for the file that kernelapp creates, just so that its distinguishable from the "native" connection files - as this might be something an admin would appreciate.

Reply:
I don't think we have a public kernel ID for arbitrary kernels at the moment. It's probably a good idea, but it also intersects with the question of restarting kernels. Should a restarted kernel have the same ID as the original? What about if you can checkpoint/restore a kernel process - should the restored process have the same ID? What about if you restore the same checkpoint to two processes?

Maybe another fudge to make things work until we think about that would be for KernelApp to check if the manager has already written a connection file and only write one if it can't use that. I don't much like it, but maybe it's better than the other bad options.

There's a bigger question about discovering running kernel instances (as distinct from discovering kernel *types*). Who should be able to discover what kernels? What part is responsible for advertising them? When should kernels be run without being advertised?

Reply:
Thinking about this, I'm inclined to leave it as is (with two connection files) for now. It's a bit messy, but it avoids coupling `jupyter-kernel` to the default providers which happen to create a local connection file.

Creating multiple connection files only matters if something wants to list all running kernels. Most use cases that I'm aware of for connection files are using the `--existing` flag, either with the name of a specific connection file, or with no name to get the most recently started kernel. As far as I know, this should still work. And if there is code listing all running kernels, it should be easy enough to adapt it to deduplicate them.

Reply:
> Should a restarted kernel have the same ID as the original? 

I think we must retain the kernel id on restarts.  The MappingKernelManager relies on this fact - otherwise it would remove the current slot and create a new slot for the new kernel manager.  In addition, I believe kernel providers should be able to do what they want within their kernel manager, but the kernel-id, IMO, is the public _key_ for identifying a kernel throughout its lifetime (where restarts are considered within its lifetime). 

I think I mentioned this before, but I believe `KernelFinder.launch()` should take an optional `kernel_id=None` parameter.  This parameter would also be honored on initial launch as well and the restarter would then use it during restarts: `self.kernel_finder.launch(self.kernel_type, cwd, kernel_id=self.kernel_manager.kernel_id)`.  (We'd want to extend KernelManager's initializer in a similar manner - with an optional kernel_id parameter.)

> What about if you can checkpoint/restore a kernel process - should the restored process have the same ID? What about if you restore the same checkpoint to two processes?

If these questions are directed at persisted sessions that get "revived" I think the id of the process is orthogonal to the id of the kernel.  When/how is process id used by clients?

In EG, when a persisted session is "revived", for example if the server came down, and another started, EG does not restart the kernel process, but instead, re-establishes connection to the kernel process.  It can do this because the kernel is remote and is (probably) still running. 

If we want the server to be able to support active/active scenarios (which we plan to do for EG), retaining the kernel's id across restarts is paramount to that - otherwise restarts would require communicating the old and new kernel ids to other servers, etc.  and I don't think we want to go there - especially since the current behavior is to retain the id.

> Who should be able to discover what kernels? What part is responsible for advertising them? When should kernels be run without being advertised?

I guess I'm not familiar enough with the direct kernel application approach.  For applications like notebook and jupyter-server, the list of running kernels should come directly from the MappingKernelManager.  Once (well, if) we add the notion of users and roles, we could then apply filtering on the results of the `/api/kernels` request.